### PR TITLE
[backup] manifest and controllers

### PR DESCRIPTION
**Pull request by msmouse**
## Motivation

Implementing state snapshot backup manifest.
Extracting backup and restore controllers for the state snapshot backup.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?
yes

## Test Plan

unit tests

### CODE REVIEW

**Comment by msmouse** [comment]
Verification of the snapshot proof will be added once we have the epoch-change-ledger-infos backup.

File: storage/backup/backup-cli/src/backup.rs [hunk 1]
```diff
@@ -63,105 +77,137 @@ impl BackupServiceClient {
...
-        if chunk_bytes.len() + size_of::<u32>() + record_bytes.len() > max_chunk_size {
-            assert!(chunk_bytes.len() <= max_chunk_size);
-            println!("Reached max_chunk_size.");
-
-            let (chunk_handle, proof_handle) = write_chunk(
-                client,
-                version,
-                storage,
+    client: Arc<BackupServiceClient>,
+    storage: Arc<dyn BackupStorage>,
```
Comment (wqfish) [inline storage/backup/backup-cli/src/backup.rs:112]: `Box`? Are these shared with something else?

File: storage/backup/backup-cli/src/manifest/state_snapshot.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+use crate::storage::FileHandle;
+use libra_crypto::HashValue;
+use libra_types::transaction::Version;
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize)]
+pub struct StateSnapshotChunk {
+    pub first_idx: usize,
+    pub last_idx: usize,
```
Comment (wqfish) [inline storage/backup/backup-cli/src/manifest/state_snapshot.rs:12]: Would be nice to have some comments o these fields.

**Comment by msmouse** [comment]
> Can we add some detailed summary to each individual commit?

Thanks. Will do.

**Pull request closed without merging**